docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,426 @@
+    * [Version Skew Strategy](#version-skew-strategy)
+* [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+
+## Summary
+
+This proposal introduces a set of Runtime Hooks designed for allowing the implementation of add-on orchestration solutions on top of Cluster API.
+
+## Motivation
+
+Cluster Resource Set (CRS) is the current add-on management tool packaged with Cluster API, but many users rely on their own package management tool like helm, kapp, ArgoCD or flux, because those tools have a broad range of capabilities that are not currently available in CRS.
+
+The ability to orchestrate add-ons in line with events in the cluster lifecycle is becoming a requirement for many CAPI users, but in order to make this possible a mechanism to plug into the cluster lifecycle is required. This proposal introduces a set of Runtime Hooks designed to meet the need for add-on management orchestration including:
+
+* Operations for installing add-ons during the cluster provisioning workflow
+* Operations for upgrading add-ons during the cluster upgrade workflow
+* Operations for handling add-ons during the cluster deletion workflow
+
+Runtime Hooks enable the Cluster Lifecycle to trigger these processes based on the current state of the cluster, allowing them to start after some state has been reached or waiting for them to complete before moving on with Cluster-wide operations such as Cluster creation or deletion.
+
+Once these hooks are in place, it will be possible to build a comprehensive add-on orchestration solution on top of Cluster API that can leverage external tools such as helm, kapp, ArgoCD, flux or eventually CRS as an alternative.
+
+The hooks defined in this proposal are going to model common events in the Cluster lifecycle and we assume that in the future they may be used for other use cases beyond add-on management. 
+
+### Goals
+
+* Identify a set of Runtime Hooks that enable management of the entire add-on lifecycle 
+* Define the OpenAPI specification of these Runtime Hooks
+* Document when the corresponding Runtime Extensions are called
+* Provide guidelines for developers implementing a corresponding Runtime Extension
+
+### Non-Goals
+
+* Defining all possible Runtime Hooks in Cluster API; this proposal defines only a subset of hooks required for add-on orchestration.
+* Define a full add-on management solution or define detailed steps for solving add-on related problems like CPI migration from in-tree to out-of-tree; this proposal is focused only in providing the foundational capabilities to do so. 
+
+### Future work
+
+* Improve the hooks defined in this proposal to better support use cases beyond add-on management 
+
+
+## Proposal
+
+This proposal adds a set of Runtime Hooks specifically designed for Cluster created from a ClusterClass and thus leveraging the idea of managed topology. 
```

Comment:
I'd expect most of the scenarios to be handled/triggered by the core operational primitives i.e Cluster/Machine CR/controller regardless of them being managed by a topology or not.

So I think before reasoning about managed topology hooks we need to clearly define the behaviour and hooks for the core Cluster and Machines lifecycle, then we can add layers atop where we think it makes sense e.g managed topology to satisfy specific use cases.

Reply:
IMO the scenarios we're talking about here can't easily be handled by the core primitives:
- The upgrade flow **BeforeClusterUpgrade**, **AfterControlPlaneUpgrade** and **AfterClusterUpgrade** has no obvious point of control outside ClusterClass.
- **BeforeClusterCreate** can't meaningfully block e.g. Machine creation for non-ClusterClass clusters.
-  **BeforeClusterDelete** or **AfterControlPlaneInitialized** could work. 

Do you have another idea on how these hooks could work without using the Topology controller as a point of control?

Reply:
>The upgrade flow BeforeClusterUpgrade, AfterControlPlaneUpgrade and AfterClusterUpgrade has no obvious point of control outside ClusterClass.

I agree that for scenarios where ordering/state between cp and Nodes or the view of both as a whole is required then topology is possibly the right lifecycle to hook into. I'm fine having this in topology, my concern is more about the fact that atm it is not possible to prove and elaborate or document "why" because the core primitives hooks are not defined anywhere, they basically don't exist and so they can't be referenced nor reason about. So I'm concern that we might not be able to make the right decisions without them being defined first.
I think at the very minimum something like your comment above should be elaborated as part of this proposal to justify the proposed hooks.

Reply:
that's a fair point.
I think that the rest of this paragraph already makes a good job in explaining the why, but it could really help if we add some examples from @killianmuldoon  comment above to make everything more concrete.
I'm making a suggestion below to address this

Reply:
I have updated the section about why the Managed Cluster were chosen for this with @fabriziopandini's suggestion. It adds more on why these hooks go into the topology controller.